blog/2019-12-18-HowToPlanAndOptimizeASecretSanta.adoc

```diff
@@ -0,0 +1,102 @@
+employees who are further apart to give gifts to each other, since they
+are the ones who probably know the least about each other. Let's optimize
+it with OptaPlanner!
+
+== The Constraints
+
+One of the most obvious constraints is that everyone should get a gift.
+Or phrased a little differently: no one should get multiple gifts.
+
+Here is how one would create this constraint in OptaPlanner using the
+constraint streams API:
+[source,java]
+----
+private Constraint sameReceiverConflict(ConstraintFactory constraintFactory) {
+    return constraintFactory
+            .fromUniquePair(SecretSantaAssignment.class,
+                    Joiners.equal(SecretSantaAssignment::getReceiver))
+            .penalize("Same Receiver", HardMediumSoftBigDecimalScore.ONE_HARD);
+}
+----
+
+Next, we need to award assignments where the gifter and the receiver
+are farther apart. Here's how we will do it with constraint streams:
+[source,java]
+----
+private Constraint largerDistanceAward(ConstraintFactory constraintFactory) {
+    return constraintFactory
+            .from(SecretSantaAssignment.class)
+            .rewardConfigurableBigDecimal("secretFactor", "Larger Distance Award",
+                              (m) -> BigDecimal.valueOf(Location.calculateDistanceBetween(m.getGifter().getLocation(),
+                                                                       m.getReceiver().getLocation())));
+}
+----
+
+Let's try solving with just these two constraints to see what we get:
+
+image::noPairChecking.png[]
+
+Alice is assigned to Daniella, Daniella is assigned to Alice...
+Bob is assigned to Austin, Austin is assigned to Bob...
+Something about this solution seems odd; it only has assignments
+where the gifter is the receiver's recipient!
+
+== What happened?
+
+This is not odd at all, but rather a consequence of how we defined
+our constraints: if Alice is the person furthest from Daniella, then
+Daniella is most likely the person furthest from Alice. This means our
+"Larger Distance Award" constraint will strongly favour pairs of employees.
+
+Is this a bad thing? Yes it is; imagine if Alice got a $100 gift from
+Daniella, but only got Daniella a $10 gift. Alice will probably feel bad,
+and Daniella will feel disappointed. If Alice gave the $10 gift to Bob
+and Bob gave a $30 gift to Daniella, Alice wouldn't feel bad and
+Daniella wouldn't feel cheated.
+
+Let's add this constraint (using, you guessed it, constraint streams!)
+[source,java]
+----
+private Constraint giftPair(ConstraintFactory constraintFactory) {
+    return constraintFactory
+            .fromUniquePair(SecretSantaAssignment.class,
+                            Joiners.equal(SecretSantaAssignment::getGifter, SecretSantaAssignment::getReceiver),
+                            Joiners.equal(SecretSantaAssignment::getReciever, SecretSantaAssignment::getGifter))
+            .penalize("Gifter-Receiver Cycle", HardMediumSoftBigDecimalScore.ONE_MEDIUM);
+}
+----
```

Comment:
The example with the two `Joiner`s is less than trivial. Maybe a short explanation of what it does, or just a link to the doc, would be helpful?

Reply:
It indeed lest than trivial; before, it was just a single Joiner on Gifter and Receiver. However, that was not the correct implementation of the constraint - with only a single Joiner, it counted the number of cycles (which in turns make it favor bigger cycles). This constraint is basically saying "All matches (a,b) where the a.gifter = b.receiver AND a.receiver = b.gifter" (in other words, a gives a present to b and b gives a present to a)